ui/src/spinbox.tsx

```diff
@@ -108,16 +111,23 @@ export const
       }, [max, min, parseValue, precision]),
       handleOnInput = React.useCallback((val: U) => {
         wave.args[name] = val
+        m.value = val
         if (trigger) wave.push()
+        // eslint-disable-next-line react-hooks/exhaustive-deps
       }, [name, trigger]),
       debouncedHandleOnInput = React.useRef(wave.debounce(DEBOUNCE_TIMEOUT, handleOnInput)),
       onInput = (e: React.SyntheticEvent<HTMLElement>) => {
         const numVal = handleValue((e.target as HTMLInputElement).value)
         trigger ? debouncedHandleOnInput.current(numVal) : handleOnInput(numVal)
       }
 
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    React.useEffect(() => { wave.args[name] = (value < min) ? min : ((value > max) ? max : value) }, [])
+    React.useEffect(() => {
+      if (value === wave.args[name]) return
```

Comment:
To prevent from precision being ignored (see the failing test case when the condition is removed). However I've just found out, this solution is not ideal since it causes other problems when using numbers with decimal point. I'm looking for another solution, will keep you updated.